rpc/grpc/api.go

```diff
@@ -37,3 +41,149 @@ func (bapi *broadcastAPI) BroadcastTx(ctx context.Context, req *RequestBroadcast
 		},
 	}, nil
 }
+
+type blockAPI struct {
+	sync.Mutex
+	heightListeners map[chan int64]struct{}
+}
+
+func newBlockAPI() *blockAPI {
+	return &blockAPI{
+		// TODO(rach-id) make 1000 configurable if there is a need for it
+		heightListeners: make(map[chan int64]struct{}, 1000),
+	}
+}
+
+func (blockAPI *blockAPI) listenForHeights() {
+	env := core.GetEnvironment()
+	newBlocksChannel, err := env.EventBus.Subscribe(context.Background(), "grpc-listener", types2.EventQueryNewBlock, 500)
+	if err != nil {
+		log.Fatalf("Failed to subscribe to new blocks: %v", err)
+	}
+	for {
+		select {
+		case <-newBlocksChannel.Cancelled():
+			env.Logger.Error("cancelled grpc subscription")
+			// TODO(rach-id): maybe retry to connect if users want this functionality
+			return
+		case event := <-newBlocksChannel.Out():
```

Comment:
Would it be possible for either channels to be closed? Without looking into implementation, it might be worth to add handling for closed channels reads to prevent potential panics

Reply:
[a490a04](https://github.com/celestiaorg/celestia-core/pull/1513/commits/a490a04614145d596b04f23ea4ea6ba9a4506604)